py/objboundmeth.c

```diff
@@ -35,6 +35,8 @@ typedef struct _mp_obj_bound_meth_t {
     mp_obj_t self;
 } mp_obj_bound_meth_t;
 
+STATIC const mp_obj_type_t mp_type_bound_meth;
```

Comment:
I'd avoid such a forward declaration of static variable (which is also a "tentative definition" !). If for some reason the actual definition of `mp_type_bound_meth` is renamed, then this line will turn into an actual definition of a (zero initialized) `mp_obj_type_t` object, with some code then potentially referring to it instead of the renamed (actual) definition.

Forward declarations of functions will always only be declarations, so moving the actual `mp_type_bound_meth` defintion up here and adding the required function declarations would IMO be better "defensive coding".

Reply:
I generally try to prevent forward declerations as much as possible, however in this case there needs to be a forward declaration of some type. Because `mp_type_bound_meth` is used in `bound_meth_binary_op` and `bound_meth_binary_op` is used in `mp_type_bound_meth`.

The only way this can be avoided is if it is not needed to ensure the type is `mp_type_bound_meth` in the `bound_meth_binary_op`. I am not 100% certain this is needed, but I'd rather be sure and check the type then risk some weird error later.

The alternative of a forward decleration/prototype of `mp_type_bound_meth` is to define a prototype for `bound_meth_binary_op`. Which is the same issue, but for a different item. But that is a change I could make.

Reply:
In addition, I choose to have a forward declaration for the object instead of the function because that pattern is used in other places in the codebase. Like `modure` https://github.com/micropython/micropython/blob/master/extmod/modure.c#L58 which is the closest comparison, as in that module the forward declaration is also used to for the `mp_obj_is_type` comparison.

Reply:
> I generally try to prevent forward declerations as much as possible, however in this case there needs to be a forward declaration of some type.

Yes, definitely.

> rather be sure and check the type

Of course, the check for the type adds runtime safety.

> The alternative of a forward decleration/prototype of `mp_type_bound_meth` is to define a prototype for `bound_meth_binary_op`. Which is the same issue, but for a different item.

Yes, that's what I was proposing: Making a "function prototype" rather than a "variable prototype" because it is safer - in terms of possible coding mistakes. A "function prototype" will never be a definition, a "variable prototype" can silently become a variable definition, therefore mask compile/link errors and cause relatively difficult to find bugs.

> that pattern is used in other places in the codebase

I (personaly) would rather change these other places, but that's just my opinion :)

Reply:
(To illustrate what kind of error I'm talking about: https://ideone.com/IaABNq)

Reply:
Hmmm.. Alright, fair point, I'll have made the change.

Reply:
Actually, we do already have quite a few instances of static local types, eg `mp_type_iobase`.  So for consistency please keep it that way (forward decl of the `mp_obj_type_t`).